projects/plugins/boost/app/modules/lazy-images/class-lazy-images.php

```diff
@@ -1,149 +0,0 @@
-<?php
-/**
- * Implements the Lazy Images feature.
- *
- * @link       https://automattic.com
- * @since      1.0.0
- * @package    automattic/jetpack-boost
- */
-
-namespace Automattic\Jetpack_Boost\Modules\Lazy_Images;
-
-use Automattic\Jetpack\Jetpack_Lazy_Images;
-use Automattic\Jetpack_Boost\Modules\Module;
-use Jetpack;
-
-/**
- * Class Lazy_Images
- */
-class Lazy_Images extends Module {
-
-	const MODULE_SLUG = 'lazy-images';
-
-	/**
-	 * Jetpack Boost plugin instance.
-	 *
-	 * @var \Automattic\Jetpack_Boost\Jetpack_Boost
-	 */
-	protected $jetpack_boost_plugin;
-
-	/**
-	 * Constructor.
-	 */
-	public function __construct() {
-		parent::__construct();
-
-		add_action( 'jetpack_boost_loaded', array( $this, 'set_jetpack_boost_plugin_instance' ), 10, 1 );
-		add_action( 'jetpack_loaded', array( $this, 'enable_jetpack_lazy_images_module' ) );
-		add_filter( 'jetpack_boost_module_enabled', array( $this, 'set_default_module_status' ), 10, 2 );
-		add_action( 'jetpack_boost_pre_set_module_status', array( $this, 'set_jetpack_lazy_images_module_status' ), 10, 2 );
-		add_action( 'jetpack_boost_pre_is_module_enabled', array( $this, 'enable_module' ), 10, 2 );
-		add_action( 'jetpack_activate_module', array( $this, 'action_jetpack_activate_module' ), 10, 2 );
-		add_action( 'jetpack_deactivate_module', array( $this, 'action_jetpack_deactivate_module' ), 10, 2 );
-	}
-
-	/**
-	 * Check whether the Jetpack Lazy Images module is enabled.
-	 */
-	public static function is_jetpack_lazy_images_module_enabled() {
-		return class_exists( 'Jetpack' ) && Jetpack::is_module_active( self::MODULE_SLUG );
```

Comment:
Ok I just imported my changes from my fixes branch, and there are two tests failing. It is not keeping the "lazy-images" feature in lock-step between Jetpack and Boost. This line is why -- the Modules used to be able to override the default status behavior.

I started down the road to fix this -- I introduced a new `JetpackDeferredStatus` class which inherited from `Status` and override its `is_enabled` and `update` methods to defer to Jetpack's module status.

But then I wasn't sure about where to go next.

I could see a few solutions, but the least kludgey to me would be:
1. To change `Feature` from an interface to a base class, so that it can have default behaviour,
2. Add a `create_status_manager` method to the `Feature` base class, which returns a new `Status` object, and override it for lazy images to return a new `JetpackDeferredStatus` object instead.
3. Profit

But that's a big enough design change that I didn't want to do it without consulting with @pyronaur first.

Alternately, we could:
- Add a `create_status_manager` method to every feature. That's kind of tiresome, or
- Add the deferral logic to the `Status` class, and have it know a bunch of slugs to defer to Jetpack's status. That feels like it puts the intelligence in the wrong place.
- Add a special case to the `Optimization` class and have _it_ know which Features need special status treatment.

None of those options seem clean to me.

So @pyronaur, where will we add this logic?

Reply:
> I started down the road to fix this -- I created a new JetpackDeferredStatus class which inherits from Status and overrides its is_enabled and update methods to defer to Jetpack's module status.

This is certainly one way of doing things. Although, whenever possible - I think we should try to think how we can do this with composition first (instead of inhertiance)

> But then how do we make some Optimizations use Status, and others use JetpackDeferredStatus?

Not entirely sure either, but I think it's worth exploring adding a status contract where both Jetpack_Status and Boost_Status both subscribe to that contract and become interchangeable. At that point, whoever is instantiating a status can decide which type of status is needed (for example, depending on whether Jetpack is active)

But there's a problem with this approach still - because both plugins can be activated without the other, so if the place of storage changes, there's no easy way to silently sync and migrate the option.

I have to look closer at the lazy images module - but I think maybe that module itself should be able to know whether it's active or not? I'm not sure, but it would be a way to neatly fix this.



> **Alternately, we could:**

Hm. I need to do a little bit of thinking on this, but it seems that if we use any type of inheritance it's automatically going to tie down the parent class.

Thinking from a single-responsibility angle - what's the single responsibility we're missing here? Sounds like we're talking about where to store the status - either in Jetpack options or Boost options. So I think that's the hint of the direction we need to go in.

No other classes should be aware that we have a compatibility layer with Jetpack. Where to place it exactly - I'm not entirely sure. I think it needs a bit of thought as a part of #22542